### PR TITLE
Update firmware attributes for SP/SPT

### PR DESCRIPTION
**Pull request by yuvirani**
### Description
Update firmware attributes for SP/SPT

### Check List
- [ ] New functionality includes testing.
  - [ ]  All tests pass for go 1.11 + gofmt checks.
- [ ] New functionality has been documented in the README if applicable.
  - [ ] New functionality has been thoroughly documented in the examples (please include helpful comments).
  - [ ] New endpoints supported are updated in the endpoints-support.md file.
- [ ] Changes are documented in the CHANGELOG.

### CODE REVIEW

File: oneview/resource_server_profile_template.go [hunk 1]
```diff
@@ -1541,12 +1533,11 @@ func resourceServerProfileTemplateCreate(d *schema.ResourceData, meta interface{
...
 			firmwareItem := raw.(map[string]interface{})
 			firmware = ov.FirmwareOption{
-				ForceInstallFirmware:     firmwareItem["force_install_firmware"].(bool),
-				FirmwareBaselineUri:      utils.NewNstring(firmwareItem["firmware_baseline_uri"].(string)),
-				ManageFirmware:           firmwareItem["manage_firmware"].(bool),
-				FirmwareInstallType:      firmwareItem["firmware_install_type"].(string),
-				FirmwareActivationType:   firmwareItem["firmware_activation_type"].(string),
-				FirmwareScheduleDateTime: firmwareItem["firmware_schedule_date_time"].(string),
+				ForceInstallFirmware:   firmwareItem["force_install_firmware"].(bool),
+				FirmwareBaselineUri:    utils.NewNstring(firmwareItem["firmware_baseline_uri"].(string)),
```
Comment (AvijitOS) [inline oneview/resource_server_profile_template.go:1537]: why we are not taking care of compliance_control property? its there in the schema

File: oneview/resource_server_profile_template.go [hunk 2]
```diff
@@ -1923,6 +1914,19 @@ func resourceServerProfileTemplateRead(d *schema.ResourceData, meta interface{})
...
 
+	emptyFirmware := ov.FirmwareOption{}
+	if spt.Firmware != emptyFirmware {
+		firmware := make([]map[string]interface{}, 0, 1)
+		firmware = append(firmware, map[string]interface{}{
+			"compliance_control":       spt.Firmware.ComplianceControl,
+			"firmware_baseline_uri":    spt.Firmware.FirmwareBaselineUri,
+			"manage_firmware":          spt.Firmware.ManageFirmware,
+			"firmware_install_type":    spt.Firmware.FirmwareInstallType,
+			"firmware_activation_type": spt.Firmware.FirmwareActivationType,
```
Comment (AvijitOS) [inline oneview/resource_server_profile_template.go:1926]: not seeing anything related to force_install_firmware property?

File: oneview/resource_server_profile_template.go [hunk 3]
```diff
@@ -2839,12 +2843,11 @@ func resourceServerProfileTemplateUpdate(d *schema.ResourceData, meta interface{
 		for _, raw := range rawFirmware {
 			firmwareItem := raw.(map[string]interface{})
 			firmware = ov.FirmwareOption{
-				ForceInstallFirmware:     firmwareItem["force_install_firmware"].(bool),
-				FirmwareBaselineUri:      utils.NewNstring(firmwareItem["firmware_baseline_uri"].(string)),
-				ManageFirmware:           firmwareItem["manage_firmware"].(bool),
-				FirmwareInstallType:      firmwareItem["firmware_install_type"].(string),
-				FirmwareActivationType:   firmwareItem["firmware_activation_type"].(string),
-				FirmwareScheduleDateTime: firmwareItem["firmware_schedule_date_time"].(string),
+				ForceInstallFirmware:   firmwareItem["force_install_firmware"].(bool),
```
Comment (AvijitOS) [inline oneview/resource_server_profile_template.go:2847]: not seeing anything related to compliance_control?

File: oneview/resource_server_profile.go [hunk 4]
```diff
@@ -2161,6 +2165,32 @@ func resourceServerProfileRead(d *schema.ResourceData, meta interface{}) error {
...
+		firmwareAttr["consistency_state"] = serverProfile.Firmware.ConsistencyState
+		firmwareAttr["reapply_state"] = serverProfile.Firmware.ReapplyState
+		firmwareAttr["firmware_baseline_uri"] = serverProfile.Firmware.FirmwareBaselineUri
+		firmwareAttr["manage_firmware"] = serverProfile.Firmware.ManageFirmware
+		firmwareAttr["firmware_install_type"] = serverProfile.Firmware.FirmwareInstallType
+		firmwareAttr["firmware_activation_type"] = serverProfile.Firmware.FirmwareActivationType
+		firmwareAttr["firmware_schedule_date_time"] = serverProfile.Firmware.FirmwareScheduleDateTime
+
+		//flattening firmware to update force_install_firmware
+		if fw, ok := d.GetOk("firmware"); ok {
```
Comment (AvijitOS) [inline oneview/resource_server_profile.go:2181]: is this mean, when firmware block is not set force_install_firmware property will be set to false.
I believe if firmware block is not set then manage_firmware also come as false. If we want to do this then spt implementation is not consistent. Need more explanation regarding the same

File: oneview/resource_server_profile_template.go [hunk 5]
```diff
@@ -1541,12 +1533,11 @@ func resourceServerProfileTemplateCreate(d *schema.ResourceData, meta interface{
...
+				FirmwareBaselineUri:    utils.NewNstring(firmwareItem["firmware_baseline_uri"].(string)),
+				ManageFirmware:         firmwareItem["manage_firmware"].(bool),
+				FirmwareInstallType:    firmwareItem["firmware_install_type"].(string),
+				FirmwareActivationType: firmwareItem["firmware_activation_type"].(string),
 			}
```
Comment (AvijitOS) [inline oneview/resource_server_profile_template.go:1541]: Have we verified the case where profile is created from template and template contains firmware setting..what happen to those properties which are not exist in template but there in sp...hope it does not create any idempotency issues

File: oneview/resource_server_profile.go (hunk 4 above)
Comment (yuvirani) [reply oneview/resource_server_profile.go:2181]: This firmware block check is only for force_install_firmware attribute.. As you can see, others attributes get updated irrespective of user config. As I mentioned above, SPT doesn't have any issues with force_install_firmware param.

File: oneview/resource_server_profile_template.go (hunk 5 above)
Comment (yuvirani) [reply oneview/resource_server_profile_template.go:1541]: Yes, I have verified them. When attribute values change from SPT to SP, we dont get any idempotency issues in TF where as you can see the warning in OV saying SP is inconsistent with SPT.

**Review by AvijitOS — APPROVED** [review]

**Review by AsisBagga — APPROVED** [review]
LGTM

**Pull request merged**